File: NUnit.vb.Test/Examples/ExceptionTest.vb
Option Explicit On

Imports NUnit.Framework

Namespace Examples

	<TestFixture>
	Public Class ExceptionTest

		<Test>
		Public Sub Excpetion_Assert_Throws_TestSub()
			Assert.Throws(Of ArgumentException)(Sub() SubThatThrows())
		End Sub

		<Test>
		Public Sub Excpetion_Assert_That_TestSub()
			Assert.That(AddressOf SubThatThrows, Throws.ArgumentException)
		End Sub
		<Test>
		Public Sub Excpetion_Assert_Throws_SubWithArgument()
			Assert.Throws(Of ArgumentException)(Sub() SubThatThrowsWithArgument(True))
		End Sub

		<Test>
		Public Sub Excpetion_Assert_That_SubWithArgument()
			Assert.That(Sub() SubThatThrowsWithArgument(True), Throws.ArgumentException)
		End Sub

		<Test>
		Public Sub Excpetion_Assert_Throws_Function()
			Assert.Throws(Of ArgumentException)(Function() FunctionThatThrows())
		End Sub

		<Test>
		Public Sub Excpetion_Assert_That_Function()
			Assert.That(AddressOf FunctionThatThrows, Throws.ArgumentException)
		End Sub

		<Test>
		Public Sub Excpetion_Assert_Throws_FunctionWithArgument()
			Assert.Throws(Of ArgumentException)(Function() FunctionThatThrowsWithArgument(True))
		End Sub

		<Test>
		Public Sub Excpetion_Assert_That_FunctionWithArgument()
			Assert.That(Sub() FunctionThatThrowsWithArgument(True), Throws.ArgumentException)
		End Sub

#Region "Infrasctructure for test"

		Private Shared Sub SubThatThrows()
			Throw New ArgumentException("Throws for test")
		End Sub

		Private Shared Sub SubThatThrowsWithArgument(ByVal shouldThrow As Boolean)
			If shouldThrow Then
				Throw New ArgumentException("Throws for test")
			End If
		End Sub

		Private Shared Function FunctionThatThrows()
			Throw New ArgumentException("Throws for test")
		End Function

		Private Shared Function FunctionThatThrowsWithArgument(ByVal shouldThrow As Boolean)
			If shouldThrow Then
				Throw New ArgumentException("Throws for test")
			Else
				Return Nothing
			End If
		End Function

#End Region

	End Class

End Namespace

File: Nunit.Test/Examples/ExceptionTest.cs
using System;
using NUnit.Framework;

namespace Nunit.Test.Examples
{
	[TestFixture]
	public class ExceptionTest
	{

		#region "Assert.Throws"

		[Test]
		public void Excpetion_Assert_Throws_TestVoid()
		{
			Assert.Throws<ArgumentException>(VoidThatThrows);
		}

		[Test]
		public void Excpetion_Assert_Throws_TestFunction()
		{
			Assert.Throws<ArgumentException>(() => FunctionThatThrows());
		}

		#endregion

		#region "Assert.That"

		[Test]
		public void Excpetion_Assert_That_TestVoid()
		{
			Assert.That(VoidThatThrows, Throws.ArgumentException);
			Assert.That(VoidThatDontThrow, Throws.Nothing);
		}

		[Test]
		public void Excpetion_Assert_That_TestFunction()
		{
			Assert.That(FunctionThatThrows, Throws.ArgumentException);
			Assert.That(FunctionThatDontThrow, Throws.Nothing);
		}

		#endregion

		#region "Assert.DoesNotThrow"
		[Test]
		public void Excpetion_Assert_DoesNotThrow_TestVoid()
		{
			Assert.DoesNotThrow(VoidThatDontThrow);
		}

		[Test]
		public void Excpetion_Assert_DoesNotThrow_TestFunction()
		{
			Assert.DoesNotThrow(() => FunctionThatDontThrow());
		}
		#endregion

		#region "Infrastructure for test"

		private static void VoidThatThrows()
		{
			throw new ArgumentException("Throws for test");
		}

		private static string FunctionThatThrows()
		{
			throw new ArgumentException("Throws for test");
		}

		private static void VoidThatDontThrow()
		{
		}

		private static string FunctionThatDontThrow()
		{
			return null;
		}

		#endregion

	}
}